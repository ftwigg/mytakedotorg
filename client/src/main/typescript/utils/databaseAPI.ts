import database, {
  CaptionMeta,
  CaptionWord,
  Database,
  VideoFact,
  User,
  Article
} from "./databaseData";
import { Foundation } from "../java2ts/Foundation";
import { Routes } from "../java2ts/Routes";
import { slugify } from "./functions";
import { TakeBlock } from "../components/BlockEditor";

export function getAllVideoFacts(): VideoFact[] {
  return database.videos;
}

export function getVideoFact(videoId: string): VideoFact {
  for (let video of database.videos) {
    if (video.id === videoId) {
      return video;
    }
  }

  throw "Cannot get VideoFact for video with id: " + videoId;
}

export function getVideoFactTitle(videoId: string): string {
  let video = getVideoFact(videoId);
  if (video) {
    return video.title;
  }

  throw "Cannot get video title for video with id: " + videoId;
}

export function getVideoFactPrimaryDate(videoId: string): Date {
  let video = getVideoFact(videoId);
  if (video) {
    return video.primaryDate;
  }

  throw "Cannot get video primary date for video with id: " + videoId;
}

export function getVideoFactPrimaryDateKind(videoId: string): string {
  let video = getVideoFact(videoId);
  if (video) {
    return video.primaryDateKind;
  }

  throw "Cannot get video primary date kind for video with id: " + videoId;
}

export function getVideoFactCaptionFile(videoId: string): string {
  let video = getVideoFact(videoId);
  if (video && video.captionFile) {
    return video.captionFile;
  }

  throw "Cannot get video caption file for video with id: " + videoId;
}

export function getVideoCaptionMetaData(videoId: string): CaptionMeta {
  let video = getVideoFact(videoId);
  if (video && video.captionMeta) {
    return video.captionMeta;
  }

  throw "Cannot get video caption metadata for video with id: " + videoId;
}

export function getVideoCaptionWordMap(videoId: string): CaptionWord[] {
  const captionFile = getVideoFactCaptionFile(videoId);
  const source = require("../foundation/" + captionFile);
  if (source) {
    return <CaptionWord[]>JSON.parse(source);
  }

  throw "Cannot get video caption word-map for video with id: " + videoId;
}

export function getAllFacts(
  callback: (
    error: string | Error | null,
    documents: Foundation.FactLink[]
  ) => any
): void {
  const headers = new Headers();

  headers.append("Accept", "application/json"); // This one is enough for GET requests

  const request: RequestInit = {
    method: "GET",
    headers: headers,
    cache: "default"
  };

  fetch(Routes.FOUNDATION_DATA_INDEX, request)
    .then(function(response: Response) {
      const contentType = response.headers.get("content-type");
      if (
        contentType &&
        contentType.indexOf("application/json") >= 0 &&
        response.ok
      ) {
        return response.json();
      } else {
        callback("Error retrieving Document Facts", []);
      }
    })
    .then(function(json: any) {
      callback(null, json);
    })
    .catch(function(error: Error) {
      // Network error
      callback(error, []);
    });
}

export function getDocumentFact(
  factHash: string,
  callback: (
    error: string | Error | null,
    documentFact: Foundation.DocumentFactContent | null
  ) => any
): void {
  const headers = new Headers();

  headers.append("Accept", "application/json"); // This one is enough for GET requests

  const request: RequestInit = {
    method: "GET",
    headers: headers,
    cache: "default"
  };

  fetch(Routes.FOUNDATION_DATA + "/" + factHash + ".json", request)
    .then(function(response: Response) {
      const contentType = response.headers.get("content-type");
      if (
        contentType &&
        contentType.indexOf("application/json") >= 0 &&
        response.ok
      ) {
        return response.json();
      } else {
        callback("Error retrieving Document Fact", null);
      }
    })
    .then(function(json: any) {
      callback(null, json);
    })
    .catch(function(error: Error) {
      // Network error
      callback(error, null);
    });
}

export function getDocumentFactTitle(excerptId: string): string {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.title;
    }
  }

  throw "Cannot get document title for document with id: " + excerptId;
}

export function getDocumentFactFilename(excerptId: string): string {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.filename;
    }
  }

  throw "Cannot get document filename for document with id: " + excerptId;
}

export function getDocumentFactPrimaryDate(excerptId: string): Date {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.primaryDate;
    }
  }

  throw "Cannot get document primary date for document with id: " + excerptId;
}

export function getDocumentFactPrimaryDateKind(excerptId: string): string {
  for (let excerpt of database.excerpts) {
    if (slugify(excerpt.title) === excerptId) {
      return excerpt.primaryDateKind;
    }
  }

  throw "Cannot get document primary date kind for document with id: " +
    excerptId;
}

export function getUser(username: string): User | null {
  for (let user of database.users) {
    if (user.name === username) {
      return user;
    }
  }

  throw "Cannot get user: " + username;
}

export function getAllUsers(): User[] {
  return database.users;
}

export function createUser(username: string): boolean {
  if (getUser(username)) {
    throw "User already exists with username: " + username;
  }

  return database.users.push({ name: username, articles: null }) > 0;
}

export function updateUserName(
  oldUserName: string,
  newUserName: string
): boolean {
  let oldUser = getUser(oldUserName);
  let newUser = getUser(newUserName);
  if (!oldUser || newUser) {
    throw "Failed to update username";
  }

  oldUser.name = newUserName;

  return true;
}

export function deleteUser(username: string): boolean {
  if (!getUser(username)) {
    throw "User doesn't exist: " + username;
  }

  let newDatabase: Database = {
    ...database,
    users: database.users.filter(user => {
      return user.name !== username;
    })
  };

  return true;
}

export function getAllArticles(username?: string): Article[] | null {
  if (username) {
    let user = getUser(username);
    if (user) {
      return user.articles;
    } else {
      return null;
    }
  }

  let articles: Article[] = [];
  for (let user of database.users) {
    if (user.articles) {
      articles = [...articles, ...user.articles];
    }
  }

  return articles;
}

export function getArticle(
  username: string,
  titleSlug: string
): Article | null {
  let user = getUser(username);
  if (user && user.articles) {
    for (let article of user.articles) {
      if (article.titleSlug === titleSlug) {
        return article;
      }
    }
  } else {
    //Invalid username
    throw "User doesn't exist: " + username;
  }

  //Article not found
  return null;
}

export function createArticle(
  username: string,
  title: string,
  blocks: TakeBlock[],
  previewBlocks: number[]
): boolean {
  if (!title) {
    // Throw error, title is required
    throw "Invalid title: " + title;
  }
  if (getArticle(username, slugify(title))) {
    // Can't create duplicate article
    throw "User, " + username + ", already has article with title, " + title;
    /**
		 * Could also create the article anyway and append an ID counter to the title,
		 * e.g. new-take-title, new-take-title-2, new-take-title-3
		 */
  }
  let user = getUser(username);
  if (user) {
    let articles = getAllArticles(username);
    if (articles) {
      articles = [
        ...articles,
        {
          title: title,
          titleSlug: slugify(title),
          blocks: blocks,
          previewBlocks: previewBlocks
        }
      ];
    } else {
      articles = [
        {
          title: title,
          titleSlug: slugify(title),
          blocks: blocks,
          previewBlocks: previewBlocks
        }
      ];
    }
    return true;
  } else {
    // Invalid username
    throw "User doesn't exist: " + username;
  }
}

export function updateArticleTitle(
  username: string,
  oldTitleSlug: string,
  newTitle: string
): boolean {
  let newArticle = getArticle(username, slugify(newTitle));
  if (newArticle) {
    // Can't create duplicate article
    throw "User, " + username + ", already has article with title, " + newTitle;
    /**
		 * Could also create the article anyway and append an ID counter to the title,
		 * e.g. new-take-title, new-take-title-2, new-take-title-3
		 */
  }

  let oldArticle = getArticle(username, oldTitleSlug);
  if (oldArticle) {
    oldArticle.title = newTitle;
    oldArticle.titleSlug = slugify(newTitle);

    return true;
  }

  throw "Failed to update article title";
}

export function updateArticle(
  username: string,
  titleSlug: string,
  newBlocks: TakeBlock[]
): boolean {
  let article = getArticle(username, titleSlug);
  if (article) {
    article.blocks = newBlocks;

    return true;
  }

  throw "Failed to update article";
}

export function deleteArticle(username: string, titleSlug: string): boolean {
  let user = getUser(username);
  if (user && user.articles) {
    let articles = user.articles.filter(article => {
      return article.titleSlug !== titleSlug;
    });
    let newUser: User = {
      ...user,
      articles: articles
    };
    let newDatabase: Database = {
      ...database,
      users: [...database.users, newUser]
    };
    return true;
  } else {
    // User or articles don't exist
    throw "Failed to delete article";
  }
}
