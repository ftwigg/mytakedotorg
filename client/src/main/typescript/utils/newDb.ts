export const documents = [
  {
    title: "United States Constitution",
    primaryDate: "1788-06-21",
    primaryDateKind: "ratified",
    kind: "document",
    url: "gD78kfjJVp5yKdVVDeDeZL03ViA4UacyOvXZFzHnQaY="
  },
  {
    title: "Bill of Rights",
    primaryDate: "1791-12-15",
    primaryDateKind: "ratified",
    kind: "document",
    url: "uvzrf9zoJssiqWJw7jPvk8O_o3uQcO4WV-zyVD6EQlE="
  },
  {
    title: "Amendment 11",
    primaryDate: "1795-02-07",
    primaryDateKind: "ratified",
    kind: "document",
    url: "5NYBgn8DlGfSS3UxSpe4_BxmpFlnS84DOwoL9ZaB7k4="
  },
  {
    title: "Amendment 12",
    primaryDate: "1804-06-15",
    primaryDateKind: "ratified",
    kind: "document",
    url: "HHLcwaOEUWBdiww5uqEVBAQ9bGA9bV94cQbg5pBEonc="
  },
  {
    title: "Amendment 13",
    primaryDate: "1865-12-06",
    primaryDateKind: "ratified",
    kind: "document",
    url: "anw0aNyzDzXJIMFc3Pdx21KLwjFWS-xe1PKT7sdT5tY="
  },
  {
    title: "Amendment 14",
    primaryDate: "1868-07-09",
    primaryDateKind: "ratified",
    kind: "document",
    url: "EoKRQLGeFGCfoeRPqPVuFhaMj1QLbnfWfHmzhUGNwLU="
  },
  {
    title: "Amendment 15",
    primaryDate: "1870-02-03",
    primaryDateKind: "ratified",
    kind: "document",
    url: "unF7vHwSF3ud3AJv_LE1CVx4oq28hWsfJAJROVPnAeE="
  },
  {
    title: "Amendment 16",
    primaryDate: "1913-02-03",
    primaryDateKind: "ratified",
    kind: "document",
    url: "LUySG57K1-SvME4eCIgqOifXtIXEi99vyiNBvx2zmXU="
  },
  {
    title: "Amendment 17",
    primaryDate: "1913-04-08",
    primaryDateKind: "ratified",
    kind: "document",
    url: "60RladvsWpRRxbcYymuRdgVzdZKYU2iKS7aSZgL-orI="
  },
  {
    title: "Amendment 18",
    primaryDate: "1919-01-16",
    primaryDateKind: "ratified",
    kind: "document",
    url: "LoEk_oxGIMLvAR8Fzo_bdSvLpMTT6yedeydxkheH3Qk="
  },
  {
    title: "Amendment 19",
    primaryDate: "1920-08-18",
    primaryDateKind: "ratified",
    kind: "document",
    url: "G8mjqCWAoAShguyukLP5XkVFnZ-wV5EWSTjKRP5iuXQ="
  },
  {
    title: "Amendment 20",
    primaryDate: "1933-01-23",
    primaryDateKind: "ratified",
    kind: "document",
    url: "IrdYwJgL2luB_2YL3H7XZREuF-zMwbquniTsgDktf0I="
  },
  {
    title: "Amendment 21",
    primaryDate: "1933-12-05",
    primaryDateKind: "ratified",
    kind: "document",
    url: "VVvhifhLz3EkpAyB8Eig_d7vjcrQBrbQtMuXhsGG9y4="
  },
  {
    title: "Amendment 22",
    primaryDate: "1951-02-27",
    primaryDateKind: "ratified",
    kind: "document",
    url: "9YXRCpeu3iHLfhp2QhkDJCUhXLvTioGMW2HU5icJiHY="
  },
  {
    title: "Amendment 23",
    primaryDate: "1961-03-29",
    primaryDateKind: "ratified",
    kind: "document",
    url: "P1jNjqVqu5df7AV53bztVSGVZ73X0GuL0bGn-r5RAK4="
  },
  {
    title: "Amendment 24",
    primaryDate: "1964-01-23",
    primaryDateKind: "ratified",
    kind: "document",
    url: "oUX4LOq2wrfbM5NvVcndK6223PvjdAh4MwMlawlaeq8="
  },
  {
    title: "Amendment 25",
    primaryDate: "1967-02-10",
    primaryDateKind: "ratified",
    kind: "document",
    url: "re_S315SZuz7D0d2PybH3TUj7a9pGMa5AT96Lz_2a7E="
  },
  {
    title: "Amendment 26",
    primaryDate: "1971-07-01",
    primaryDateKind: "ratified",
    kind: "document",
    url: "-188YXOj6s7ulIWDEqTtYwmifFR30eD3inVrky9HcP4="
  },
  {
    title: "Amendment 27",
    primaryDate: "1992-05-05",
    primaryDateKind: "ratified",
    kind: "document",
    url: "a4rvlRzn6OjfY_pWeVroPGP4oMWFDSBQwwXQh1U3pus="
  }
];
