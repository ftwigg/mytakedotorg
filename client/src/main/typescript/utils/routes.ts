export const routes = {
  DRAFTS: "/drafts/",
  DRAFTS_NEW: "/drafts/new",
  DRAFTS_DELETE: "/drafts/delete",
  DRAFTS_PUBLISH: "/drafts/publish",
  DRAFTS_SAVE: "/drafts/save",
  LOGIN: "/login",
  LOGOUT: "/logout"
};
