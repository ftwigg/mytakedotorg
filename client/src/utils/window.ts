/**
* Small screen navigation toggle
**/
function addNavEvents() {
  let body = document.body;
  let navToggle = document.getElementsByClassName("header__nav-toggle")[0];
  let navMenu = document.getElementsByClassName("nav")[0];

  function navEvent(e: Event) {
    if (!e.defaultPrevented) {
      toggleNav();
      e.preventDefault();
    }
  }

  function bodyEvent(e: Event) {
    if (body.classList.contains("fade") && !e.defaultPrevented) {
      if (
        e.srcElement &&
        !e.srcElement.classList.contains("header__nav--icon")
      ) {
        if (
          !e.srcElement.classList.contains("nav__link") &&
          !e.srcElement.classList.contains("nav__link-text")
        ) {
          toggleNav();
          e.preventDefault();
        }
      }
    }
  }

  if (navToggle) {
    navToggle.addEventListener("click", navEvent);
  }

  body.addEventListener("click", bodyEvent);
  body.addEventListener("touchstart", bodyEvent);

  function toggleNav() {
    if (navMenu) {
      if (navMenu.classList.contains("collapse")) {
        navMenu.classList.remove("collapse");
      } else {
        navMenu.classList.add("collapse");
      }
    }

    if (body.classList.contains("fade")) {
      body.classList.remove("fade");
    } else {
      body.classList.add("fade");
    }
  }
}

addNavEvents();

(window as any).YTConfig = {
  host: "https://www.youtube.com"
};

interface LoginCookie {
  username: string;
}

const cookieStr = document.cookie;
if (cookieStr) {
  const loginUi: LoginCookie = JSON.parse(
    JSON.parse(cookieStr.split("; ")[0].split("=")[1])
  );
  const loginLink = document.getElementById("login");
  if (loginLink) {
    loginLink.innerText = loginUi.username;
    loginLink.setAttribute("href", "/logout");
  }
}
