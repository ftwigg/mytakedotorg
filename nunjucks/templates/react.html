{% extends "reactCore.html" %}

{% block bundle %}
	<!-- Main -->
	<script src="/app.bundle.js"></script>
{% endblock %}
