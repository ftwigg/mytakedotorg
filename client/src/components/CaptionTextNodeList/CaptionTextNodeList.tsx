import * as React from "react";
import * as ReactDOM from "react-dom";
import { FoundationNode } from "../../utils/functions";
import { CaptionWord, CaptionMeta, SpeakerMap } from "../../utils/databaseData";
import CaptionTextNode from "../CaptionTextNode";
var bs = require("binary-search");

interface CaptionTextNodeListProps {
  className: string;
  onMouseUp: () => void;
  captionTimer: number;
  documentNodes: FoundationNode[];
  captionWordMap: CaptionWord[];
  captionMeta: CaptionMeta;
}

interface CaptionTextNodeListState {}

class CaptionTextNodeList extends React.Component<
  CaptionTextNodeListProps,
  CaptionTextNodeListState
> {
  private captionNodeContainer: HTMLDivElement;
  constructor(props: CaptionTextNodeListProps) {
    super(props);
  }
  setScrollView = () => {
    if (this.props.captionTimer) {
      const timer = this.props.captionTimer;

      let wordIdx = bs(
        this.props.captionWordMap,
        timer,
        (word: CaptionWord, time: number) => {
          return word.timestamp - time;
        }
      );

      // usually the timestamp is between two words, in which case it returns (-insertionPoint - 2)
      if (wordIdx < 0) {
        wordIdx = -wordIdx - 2;
      }

      // find the speaker for that word
      let speakerIdx = bs(
        this.props.captionMeta.speakerMap,
        wordIdx,
        (speaker: SpeakerMap, idx: number) => {
          return speaker.range[0] - idx;
        }
      );
      if (speakerIdx < 0) {
        speakerIdx = -speakerIdx - 2;
        if (speakerIdx < 0) {
          // If still negative, it means we're at the first node
          speakerIdx = 0;
        }
      }

      let speakerRange = this.props.captionMeta.speakerMap[speakerIdx];
      let hiddenTextElement;
      // This lookup should match the DOM structure of CaptionTextNode
      if (
        this.captionNodeContainer.children[speakerIdx] &&
        this.captionNodeContainer.children[speakerIdx].children[1]
      ) {
        hiddenTextElement = this.captionNodeContainer.children[speakerIdx]
          .children[1].children[0];
      } else {
        throw "Couldn't find caption node at index " + speakerIdx;
      }

      let height = 0;
      let numberOfLines = -1;
      hiddenTextElement.innerHTML = "";
      for (let i = speakerRange.range[0]; i < wordIdx; ++i) {
        hiddenTextElement.innerHTML += this.props.captionWordMap[i].word;
        if (hiddenTextElement.clientHeight !== height) {
          height = hiddenTextElement.clientHeight;
          numberOfLines++;
        }
      }

      // scroll to here
      let parentTop = this.captionNodeContainer.getBoundingClientRect().top;

      // Get the offsetTop value of the child element.
      // The line height is 25px, so add 25 for each
      // time the line has wrapped, which is equal to
      // the value of the `idx` variable
      let childTop =
        (this.captionNodeContainer.children[speakerIdx] as HTMLElement)
          .offsetTop +
        25 * numberOfLines;

      // Set the parent container's scrollTop value to the offsetTop
      this.captionNodeContainer.scrollTop = childTop;
    }
  };
  componentDidMount() {
    if (this.props.captionTimer) {
      this.setScrollView();
    }
  }
  componentDidUpdate(
    prevProps: CaptionTextNodeListProps,
    prevState: CaptionTextNodeListState
  ) {
    if (
      this.props.captionTimer &&
      prevProps.captionTimer !== this.props.captionTimer
    ) {
      this.setScrollView();
    }
  }
  render() {
    let wordCount: number;
    let nextWordCount: number;
    return (
      <div
        className={this.props.className}
        onMouseUp={this.props.onMouseUp}
        ref={(captionNodeContainer: HTMLDivElement) =>
          (this.captionNodeContainer = captionNodeContainer)}
      >
        {this.props.documentNodes.map(function(
          element: FoundationNode,
          index: number
        ) {
          return (
            <CaptionTextNode
              key={element.props.offset}
              documentNode={element}
            />
          );
        })}
      </div>
    );
  }
}

export default CaptionTextNodeList;
