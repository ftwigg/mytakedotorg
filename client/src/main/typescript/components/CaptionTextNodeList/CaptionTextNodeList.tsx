import * as React from "react";
import * as ReactDOM from "react-dom";
import { FoundationNode } from "../../utils/functions";
import { CaptionWord, CaptionMeta, SpeakerMap } from "../../utils/databaseData";
import CaptionTextNode from "../CaptionTextNode";
var bs = require("binary-search");

interface CaptionTextNodeListProps {
  className: string;
  onMouseUp: () => void;
  captionTimer: number;
  documentNodes: FoundationNode[];
  captionWordMap: CaptionWord[];
  captionMeta: CaptionMeta;
}

interface CaptionTextNodeListState {
  currentSpeaker: string;
}

class CaptionTextNodeList extends React.Component<
  CaptionTextNodeListProps,
  CaptionTextNodeListState
> {
  private captionNodeContainer: HTMLDivElement | null;
  private timerId: number | null;
  constructor(props: CaptionTextNodeListProps) {
    super(props);

    this.state = {
      currentSpeaker: "-"
    };
  }
  clearTimer = () => {
    if (this.timerId) {
      window.clearTimeout(this.timerId);
      this.timerId = null;
      // Call it one more time to be sure the result is correct when the user scroll has stopped.
      this.handleScroll();
    }
  };
  handleScroll = () => {
    // Only allow this function to execute no more than twice per second
    if (!this.timerId && this.captionNodeContainer) {
      const parentTop = this.captionNodeContainer.scrollTop;
      if (parentTop === 0) {
        this.setState({
          currentSpeaker: "-"
        });
      } else {
        let speakerIdx = bs(
          this.captionNodeContainer.children,
          parentTop,
          (child: HTMLDivElement, parentTop: number) => {
            return child.offsetTop - parentTop;
          }
        );

        if (speakerIdx < 0) {
          speakerIdx = -speakerIdx - 2;
          if (speakerIdx < 0) {
            // If still negative, it means we're at the first node
            speakerIdx = 0;
          }
        }

        this.setState({
          currentSpeaker: this.props.captionMeta.speakerMap[speakerIdx].speaker
        });

        this.timerId = window.setTimeout(this.clearTimer, 500);
      }
    }
  };
  setScrollView = () => {
    if (this.captionNodeContainer) {
      const timer = this.props.captionTimer;

      let wordIdx = bs(
        this.props.captionWordMap,
        timer,
        (word: CaptionWord, time: number) => {
          return word.timestamp - time;
        }
      );

      // usually the timestamp is between two words, in which case it returns (-insertionPoint - 2)
      if (wordIdx < 0) {
        wordIdx = -wordIdx - 2;
      }

      // find the speaker for that word
      let speakerIdx = bs(
        this.props.captionMeta.speakerMap,
        wordIdx,
        (speaker: SpeakerMap, idx: number) => {
          return speaker.range[0] - idx;
        }
      );
      if (speakerIdx < 0) {
        speakerIdx = -speakerIdx - 2;
        if (speakerIdx < 0) {
          // If still negative, it means we're at the first node
          speakerIdx = 0;
        }
      }

      let speakerRange = this.props.captionMeta.speakerMap[speakerIdx];
      const captionTextContainer = this.captionNodeContainer.children[
        speakerIdx
      ].children[1];
      let hiddenTextElement;
      // This lookup should match the DOM structure of CaptionTextNode
      if (captionTextContainer && captionTextContainer.children[1]) {
        hiddenTextElement = captionTextContainer.children[1];
      } else {
        throw "Couldn't find caption node at index " + speakerIdx;
      }

      let height = 0;
      let numberOfLines = -1;
      hiddenTextElement.innerHTML = "";
      for (let i = speakerRange.range[0]; i < wordIdx; ++i) {
        hiddenTextElement.innerHTML += this.props.captionWordMap[i].word;
        if (hiddenTextElement.clientHeight !== height) {
          height = hiddenTextElement.clientHeight;
          numberOfLines++;
        }
      }

      const speakerNameHeight = this.captionNodeContainer.children[speakerIdx]
        .children[0].clientHeight;
      let totalSpeakerHeight;
      if (speakerNameHeight === 31) {
        // Large screen, margin-bottom is 24px;
        totalSpeakerHeight = 55;
      } else {
        // Small screen, margin-bottom is 19px;
        totalSpeakerHeight = 43;
      }

      // Get the offsetTop value of the child element.
      // The line height is 25px, so add 25 for each
      // time the line has wrapped, which is equal to
      // the value of the `idx` variable.
      // Add the height and margin of the speaker name
      const childTop =
        (this.captionNodeContainer.children[speakerIdx] as HTMLElement)
          .offsetTop +
        25 * numberOfLines +
        totalSpeakerHeight;

      // Set the parent container's scrollTop value to the offsetTop
      this.captionNodeContainer.scrollTop = childTop;
    }
  };
  componentDidMount() {
    this.setScrollView();
  }
  componentDidUpdate(
    prevProps: CaptionTextNodeListProps,
    prevState: CaptionTextNodeListState
  ) {
    if (
      this.props.captionTimer &&
      prevProps.captionTimer !== this.props.captionTimer
    ) {
      this.setScrollView();
    }
  }
  render() {
    let wordCount: number;
    let nextWordCount: number;
    if (
      this.props.captionMeta.speakerMap.length !==
      this.props.documentNodes.length
    ) {
      throw "Speaker map length not equal to number of caption nodes.";
    }
    const speakerMap = this.props.captionMeta.speakerMap;
    return (
      <div className="document__text">
        <h3 className="document__node-text document__node-text--speaker-top">
          {this.state.currentSpeaker}
        </h3>
        <div
          className={this.props.className}
          onMouseUp={this.props.onMouseUp}
          ref={(captionNodeContainer: HTMLDivElement) =>
            (this.captionNodeContainer = captionNodeContainer)}
          onScroll={this.handleScroll}
        >
          {this.props.documentNodes.map(function(
            element: FoundationNode,
            index: number
          ) {
            return (
              <CaptionTextNode
                key={element.offset}
                documentNode={element}
                speaker={speakerMap[index].speaker}
              />
            );
          })}
        </div>
      </div>
    );
  }
}

export default CaptionTextNodeList;
