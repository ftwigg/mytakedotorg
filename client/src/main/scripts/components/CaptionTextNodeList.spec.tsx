import * as React from "react";
import * as renderer from "react-test-renderer";
import {} from "jest";
import CaptionTextNodeList from "./CaptionTextNodeList";
import { videoFact, videoNodes } from "../utils/testUtils";

const onMouseUp = jest.fn();
const onScroll = jest.fn();

const eventHandlers = {
  onMouseUp: onMouseUp,
  onScroll: onScroll
};
const className = "document__text document__text--caption";

jest.mock("./CaptionTextNode", () => ({
  default: "CaptionTextNode"
}));

const view = {
  start: 0,
  end: 5
};

test("CaptionTextNodeList", () => {
  const tree = renderer
    .create(
      <CaptionTextNodeList
        documentNodes={videoNodes}
        eventHandlers={eventHandlers}
        className={className}
        captionTimer={0}
        captionTranscript={videoFact.transcript}
        speakerMap={videoFact.speakerMap}
        view={view}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
