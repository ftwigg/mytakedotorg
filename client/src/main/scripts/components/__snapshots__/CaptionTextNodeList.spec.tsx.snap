// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaptionTextNodeList 1`] = `
<div
  className="document__text"
>
  <h3
    className="document__node-text document__node-text--speaker-top"
  >
    -
  </h3>
  <div
    className="document__text document__text--caption"
    onMouseUp={[MockFunction]}
    onScroll={[Function]}
  >
    <CaptionTextNode
      documentNode={
        Object {
          "component": "p",
          "innerHTML": Array [
            "Good evening I'm Martha Raddatz from ABC News.",
          ],
          "offset": 0,
        }
      }
      speaker="Raddatz"
    />
    <CaptionTextNode
      documentNode={
        Object {
          "component": "p",
          "innerHTML": Array [
            "And I'm Anderson Cooper from CNN. We want to welcome you to Washington University in St. Louis for the second presidential debate between Hillary Clinton and Donald Trump sponsored by the Commission on Presidential Debates. tonight's debate is a Town Hall format, which gives voters a chance to directly ask the candidates questions. Martha and I will ask follow-up questions but the night really belongs to the people in this room and to people across the country who have submitted questions online.",
          ],
          "offset": 46,
        }
      }
      speaker="Cooper"
    />
  </div>
</div>
`;
